File: decodificadorcesar.java
import java.util.Scanner;

public class decodificadorcesar {
    public static void main(String[] args) {
        Scanner lector = new Scanner(System.in);
        System.out.println("Ingresa el mensaje cifrado");
        String mensajeCifrado = lector.next();
        System.out.println("Cuántas veces se movió la letra al cifrar");
        int veces = lector.nextInt();

        String mensajeDescifrado = "";

        // Recorrer cada caracter de el mensaje
        for (int i = 0; i < mensajeCifrado.length(); i++) {
            char letra = mensajeCifrado.charAt(i);
            // Validar que las letras sean mayusculas o minusculas
            boolean minuscula = (letra >= 'a' && letra <= 'z');
            boolean mayuscula = (letra >= 'A' && letra <= 'Z');
            // En caso de no ser una mayuscula o minuscula se agrega al mensaje sin cambios
            if (!(minuscula || mayuscula))
                mensajeDescifrado += letra;
            else {
                // Se obtiene el numero entero de cada letra
                int ascii = (int) letra;
                // Se obtiene el numero de la primer letra en el alfabeto
                int baseAscii = (int) 'a';
                if (mayuscula)
                    baseAscii = (int) 'A';
                /* Se aplica el modulo 26 dos veces para mantener el resultad
                dentro de el rango [0,25], si solo se aplicara una vez el modulo
                26 aun asi el resultado podria ser negativo en caso de que "veces"
                sea muy grande
                 */
                int nuevoAscii = ((ascii - baseAscii - veces) % 26 + 26) % 26 + baseAscii;
                // Se convierte el valor numerico de cada letra en un caracter
                char nuevaLetra = (char) nuevoAscii;
                mensajeDescifrado += nuevaLetra;
            }
        }
        System.out.println("Mensaje descifrado: " + mensajeDescifrado);

    }
}

File: decodificadorcesar.py
# Se crea una función propia para decifrar el mensaje
# A la función se le dan los parametros necesarios
def decifrado(mensaje_cifrado, veces_caracter):
    # Se crea un string vacio para almacenar cada caracter
    mensaje_decifrado = ""
    # Se usa un ciclo for con una función range
    # A la función range solo se le pasa el parametro stop
    # Los valores default de start y step son funcionales en este caso
    for i in range(len(mensaje_cifrado)):
        # Se recorre el mensaje letra por letra en la posición i
        letra = mensaje_cifrado[i]
        # Se valida que las letras sean mayusculas o minusculas
        mayuscula = letra.isupper()
        minuscula = letra.islower()
        # Se agrega los caracteres sin cambios si no son mayusculas o minusculas
        if not (mayuscula or minuscula):
            mensaje_decifrado += letra
        else:
            # Con la funcion ord se convierte cada letra en un entero
            ascii = ord(letra)
            ascii_base = ord('A')
            if minuscula:
                ascii_base = ord('a')
                # Para descifrar el mensaje se resta el ascii
                # Menos el ascii de la primer letra del alfabeto
                # Y se le resta las veces movidas del caracter
                # Se aplica el modulo 26 para mantenerlo entre el rango del abecedario
            ascii_decifrado = (ascii - ascii_base - veces_caracter) % 26 + ascii_base
            # Se convierte los valores ascii en char
            letra_decifrada = chr(ascii_decifrado)
            mensaje_decifrado += letra_decifrada

    print(mensaje_decifrado)

decifrado(input("Ingrese el mensaje cifrado: "),int(input("Ingrese la cantidad de veces que se movio el caracter: ")))

















